### PR TITLE
SqlClient CLR UDT binary support

### PR DESCRIPTION
**Pull request by pensivebrian**
Adding SqlClient support to handle CLR UDTs as binary

@saurabh500 @YoungGah 

Currently, SqlClient on .NET Core throws an exception when it encounters a CLR UDT in the result set.

This change updates SqlClient to support working with CLR UDTs as binary data. The SqlClient API to work with UDTs as binary already exists in the full .NET Framework. This change just unblocks the API on .NET Core. The SqlClient surface area is not changed. 

The following SqlDataReader methods can be used to retrieve as UDT as binary:
- SqlDataReader.GetSqlBytes() 
- SqlDataReader.GetStream()
- SqlDataReader.GetBytes()

With this change, retrieving a CLR UDT instance is still not supported by SqlClient. The following SqlClient APIs to retrieve a CLR UDT instance have been updated to throw a PlatformNotSupported exception:
- SqlDataReader[] get property indexer 
- SqlDataReader.GetValue()
- SqlDataReader.GetValues() 
- SqlDataReader.GetSqlValue() 
- SqlDataReader.GetSqlValues()
- SqlDataReader.GetFieldValue<T>()
- SqlDataReader.GetFieldValueAsync<T>()
- SqlDataReader.GetProviderSpecificFieldType() 
- SqlDataReader.GetProviderSpecificValue() 
- SqlDataReader.GetProviderSpecificValues() 
- SqlCommand.ExecuteScalar()

For result set metadata obtained by calling SqlDataReader.GetColumnSchema(), we now populate the following properties for CLR UDTs: 
- Populate the DbColumn.UdtAssemblyQualifiedName property 
- Populate the DbColumn.DataType property with a Byte[] type

Using a Byte[] for DbColumn.DataType has interesting forward compat implications. Ideally, we'd return the Type for the underlying UDT. Since that's not possible, I feel returning a Byte[] is the best compromise as it describes how to work with the CLR UDT as binary data. If/when .NET Core ever supports CLR UDTs, we'd have to think about the impact of changing DbColumn.DataType to return the CLR UDT type.

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs [hunk 1]
```diff
@@ -1020,7 +1025,7 @@ private Type GetProviderSpecificFieldTypeInternal(_SqlMetaData metaData)
 
                 if (metaData.type == SqlDbType.Udt)
                 {
-                    throw UdtNotSupportedException();
+                    providerSpecificFieldType = MetaType.MetaMaxVarBinary.SqlType;
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: Why did you change this? `GetProviderSpecificFieldTypeInternal` is called by `GetProviderSpecificFieldType`
In `GetProviderSpecificFieldType` you are already throwing an exception on Udt.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs [hunk 2]
```diff
@@ -2051,24 +2061,8 @@ private object GetSqlValueFromSqlBufferInternal(SqlBuffer data, _SqlMetaData met
             }
             else if (_typeSystem != SqlConnectionString.TypeSystem.SQLServer2000)
             {
-                // TypeSystem.SQLServer2005
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: You have removed the check from `GetSqlValueFromSqlBufferInternal` and moved them to the callers. Why is that? 
All the callers of `GetSqlValueFromSqlBufferInternal` are throwing an exception on Udt.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs [hunk 3]
```diff
@@ -3682,17 +3682,20 @@ private bool TryCommonProcessMetaData(TdsParserStateObject stateObj, _SqlMetaDat
...
 
                 if (col.length == TdsEnums.SQL_USHORTVARMAXLEN)
                 {
                     Debug.Assert(tdsType == TdsEnums.SQLXMLTYPE ||
-                                 tdsType == TdsEnums.SQLBIGVARCHAR ||
-                                 tdsType == TdsEnums.SQLBIGVARBINARY ||
-                                 tdsType == TdsEnums.SQLNVARCHAR ||
-                                 tdsType == TdsEnums.SQLUDT,
-                                 "Invalid streaming datatype");
+                                    tdsType == TdsEnums.SQLBIGVARCHAR ||
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: Nit: Unnecessary spaces.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs [hunk 4]
```diff
@@ -472,13 +472,21 @@ internal class SqlMetaDataPriv
...
+        internal string udtAssemblyQualifiedName;
+        
         // Xml specific metadata
         internal string xmlSchemaCollectionDatabase;
         internal string xmlSchemaCollectionOwningSchema;
         internal string xmlSchemaCollectionName;
         internal MetaType metaType; // cached metaType
-
-
+
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs]: Nit: Unnecessary spaces

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs (hunk 1 above)
Comment (pensivebrian) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: I'll revert this. In my original implementation, I converted everything to use binary and missed reverting this. I'll update to throw here instead of in GetProviderSpecificFieldType.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs (hunk 2 above)
Comment (pensivebrian) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: Nice catch. I'll remove the check in the callers and have a single check here.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs (hunk 3 above)
Comment (pensivebrian) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: Will fix

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs (hunk 4 above)
Comment (pensivebrian) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs]: Will fix
